[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91044 [FSDP][7/N] Support `replicate` in `fully_shard`
* **#90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal**
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90945 fully_shard load state_dict

This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

Commit:
[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"




[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base.

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base.

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -29,6 +29,24 @@
+- ``_get_fsdp_states()`` returns all ``_FSDPState`` s in the tree.
+- ``get_fsdp_root_states()`` returns all local root ``_FSDPState`` s in the
+tree (i.e. those with ``_is_root == True``).
+- ``_get_fsdp_handles()``returns all ``FlatParamHandle`` s in the tree.
```
would I be correct that the handles returned by `_get_fsdp_handles()` and `_FSDPState` returned by `_get_fsdp_states()` should always match (come from the same fully sharded module)?

awgu replied:
Yup. `_get_fsdp_handles` just grabs the `._handles` from the return value of `_get_fsdp_states`.

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

[ghstack-poisoned]

Commit:
Update on "[FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal"


This adds a note to explain how to do traversal in the new code base. These traversal helper methods were introduced in [1/N], [3/N], and [5/N].

I am working on updating the traversal helpers to account for other composable APIs (e.g. `replicate`). The rule is that the traversal should not proceed into an incompatible API's tree. This will be needed for `fully_shard` to be above `replicate`.

[ghstack-poisoned]
